OpenRA.Mods.Cnc/Traits/World/WithResourceAnimation.cs

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common;
+using OpenRA.Mods.Common.Effects;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Cnc.Traits
+{
+	[Desc("Allows to play animations on resources.", "Attach this to the world actor.")]
+	public class WithResourceAnimationInfo : TraitInfo, Requires<IResourceLayerInfo>
+	{
+		[FieldLoader.Require]
+		[Desc("Resource types to animate.")]
+		public readonly HashSet<string> Types = null;
+
+		[Desc("The percentage of resource cells to play the animation on.", "Use two values to randomize between them.")]
+		public readonly int[] Ratio = { 5 };
+
+		[Desc("Tick interval between two animation spawning.", "Use two values to randomize between them.")]
+		public readonly int[] Interval = { 150 };
+
+		[FieldLoader.Require]
+		[Desc("Animation image.")]
+		public readonly string Image = null;
+
+		[SequenceReference(nameof(Image))]
+		[Desc("Animation sequences.")]
+		public readonly string[] Sequences = new string[] { "idle" };
+
+		[PaletteReference]
+		[Desc("Animation palette.")]
+		public readonly string Palette = null;
+
+		public override object Create(ActorInitializer init) { return new WithResourceAnimation(init.Self, this); }
+	}
+
+	class WithResourceAnimation : ITick, IWorldLoaded
+	{
+		readonly IResourceLayer resourceLayer;
+		readonly WithResourceAnimationInfo info;
+
+		readonly World world;
+		readonly HashSet<CPos> cells = new HashSet<CPos>();
+
+		int ticks;
+
+		public WithResourceAnimation(Actor self, WithResourceAnimationInfo info)
+		{
+			world = self.World;
+			this.info = info;
+
+			ticks = Common.Util.RandomDelay(self.World, info.Interval);
+
+			resourceLayer = self.Trait<IResourceLayer>();
+			resourceLayer.CellChanged += UpdateCells;
+		}
+
+		void UpdateCells(CPos cell, ResourceType resType)
+		{
+			if (resType == null)
+			{
+				cells.Remove(cell);
+				return;
+			}
+
+			if (info.Types.Contains(resType.Info.Type))
+			{
+				var resourceContent = resourceLayer.GetResource(cell);
+				if (resourceContent.Density > 0)
+					cells.Add(cell);
+				else
+					cells.Remove(cell);
+			}
+		}
+
+		void IWorldLoaded.WorldLoaded(World w, WorldRenderer wr)
+		{
+			foreach (var cell in w.Map.AllCells)
+			{
+				var type = resourceLayer.GetResource(cell).Type;
+				if (type != null && info.Types.Contains(type.Info.Type))
+					cells.Add(cell);
+			}
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (--ticks > 0)
+				return;
+
+			var twinkleable = cells.Shuffle(world.SharedRandom);
```

Comment:
nit: `twinkleable` doesn't make much sense in this more generic trait now. Perhaps something like `cellsToAnimate`.

Reply:
Yes, that is way better.